shard/shard.go

```diff
@@ -90,23 +119,40 @@ func (s shards) NumShards() int {
 	return len(s.shardsMap)
 }
 
-func (s shards) ShardIDs() []uint32 {
-	r := make([]uint32, 0, s.NumShards())
-	for s := range s.shardsMap {
-		r = append(r, s)
-	}
-	return r
+func (s shards) Shard(id uint32) (Shard, bool) {
+	shard, ok := s.shardsMap[id]
+	return shard, ok
 }
 
-func (s shards) AddShard(shard uint32) {
-	s.shardsMap[shard] = NewShard(shard)
+func (s shards) Add(shard Shard) {
+	s.shardsMap[shard.ID()] = shard
 }
 
-func (s shards) RemoveShard(shard uint32) {
+func (s shards) Remove(shard uint32) {
 	delete(s.shardsMap, shard)
 }
 
-func (s shards) ContainsShard(shard uint32) bool {
+func (s shards) Contains(shard uint32) bool {
 	_, ok := s.shardsMap[shard]
 	return ok
 }
+
+func (s shards) NumShardsInState(state State) int {
```

Comment:
Perhaps slightly more idiomatic naming here would be `NumShardsForState(...)` and `ShardsForState(...)`?